Config Syncer

xvi10 opened the pull request.

SentinelDev1 committed:
Create ConfigSyncer.sol

SentinelDev1 committed:
Update ConfigSyncer.sol

SentinelDev1 committed:
Update ConfigSyncer.sol

SentinelDev1 committed:
Update ConfigSyncer.sol

SentinelDev1 committed:
Update ConfigSyncer.sol

SentinelDev1 committed:
Add LIMITED_CONFIG_KEEPER modifier to RoleModule

For use in the ConfigSyncer contract

SentinelDev1 committed:
Update Errors.sol

SentinelDev1 committed:
Create IRiskOracle.sol

SentinelDev1 committed:
Update IRiskOracle.sol

Allows the ConfigSyncer to interact with the Chaos Labs Risk Oracle contract

SentinelDev1 committed:
Update Cast.sol

Added function used in ConfigSync to convert uint256 stored as bytes to uint256 and handle cases where the bytes is not padded to 32 bytes (as is needed to use abi.decode)

SentinelDev1 committed:
Update ConfigSyncer.sol

A number of changes including:
-Removed isParameterDisabled mapping and corresponding logic
-Moved the bytesToUint256() function to the Cast library
-Removed modifiers and thus added LIMITED_CONFIG_KEEPER modifier to RoleModule
-RoleModule is now inherited
-Removed require statements and added custom errors in the Errors contract and updated to use if statements plus custom errors
-Moved IRiskOracle to a separate file in the config folder
-Various minor optimizations including: N
-Now using keccak256(abi.encode()) for setIsParameterTypeDisabled and setIsMarketParameterDisabled mappings
-Caching market and parameterType in the sync function to local variables
-Using calldata instead of memory in function inputs
-Updated input variable naming and comment style to better reflect codebase norms

SentinelDev1 committed:
Update ConfigSyncer.sol

Minor formatting update

SentinelDev1 committed:
Update ConfigSyncer.sol

SentinelDev1 committed:
Cleanup ConfigSyncer, variables, imports and constructor

Removing unused imports, variables and updating the constructor accordingly.

SentinelDev1 committed:
Update Errors.sol

Added SyncConfig prefix to all ConfigSync contract related errors and added SyncConfigUpdatesDisabled for when all updates via ConfigSync are disabled

SentinelDev1 committed:
Update Keys.sol

Added SYNC_CONFIG_FEATURE_DISABLED, SYNC_CONFIG_LATEST_UPDATE_ID and SYNC_CONFIG_UPDATE_COMPLETED keys used to store relevant data for the ConfigSyncer contract

SentinelDev1 committed:
Update Config.sol

Added the 3 keys that were recently added to Keys.sol so that they can be updated as needed.

SentinelDev1 committed:
ConfigSyncer now using dataStore for storage

-Removed all mappings and the uint256 latestUpdate variable and now storing these data points in dataStore using keys. 

-Removed setter functions since this data will now be set via the Config contract. 

-Other minor optimizations including caching local variables where possible.

SentinelDev1 committed:
Moved local variable declarations out of for loop

Moved local variable declarations previously being done in the for loop to before the loop in order to slightly optimize gas usage.

SentinelDev1 committed:
Create RiskOracle.sol

SentinelDev1 committed:
Update RiskOracle.sol

SentinelDev1 committed:
Update RiskOracle.sol

Updated so will compile in local environment

SentinelDev1 committed:
Rename RiskOracle.sol to MockRiskOracle.sol

SentinelDev1 committed:
Update ConfigSyncer.sol

SentinelDev1 committed:
Removed error for SyncConfigUpdatesDisabled

Removed the SyncConfigUpdatesDisabled error as such cases will be handled by FeatureUtils and thus use the DisabledFeature error instead

SentinelDev1 committed:
Removed 2 keys and added 3 to allowedBaseKeys

Removed from allowedBaseKeys: SYNC_CONFIG_LATEST_UPDATE_ID and SYNC_CONFIG_UPDATE_COMPLETED as they will be updated directly in dataStore

Added to allowedBaseKeys: SYNC_CONFIG_MARKET_DISABLED, SYNC_CONFIG_PARAMETER_TYPE_DISABLED and SYNC_CONFIG_MARKET_PARAMETER_DISABLED so they can be updated via the Config contract by the keeper

SentinelDev1 committed:
Added 3 keys and 6 key getter functions

Added 3 keys: SYNC_CONFIG_MARKET_DISABLED, SYNC_CONFIG_PARAMETER_TYPE_DISABLED and SYNC_CONFIG_MARKET_PARAMETER_DISABLED rather than using SYNC_CONFIG_FEATURE_DISABLED as the base key since these are more granular and not disabling the entire feature

Added 6 functions: syncConfigFeatureDisabledKey, syncConfigMarketDisabledKey, syncConfigParameterTypeDisabledKey, syncConfigMarketParameterDisabledKey, syncConfigUpdateCompletedKey and syncConfigLatestUpdateIdKey for use in the ConfigSyncer contract

SentinelDev1 committed:
Update Keys.sol

Minor clean-up related to previous update

SentinelDev1 committed:
Various updates related to local variables and keys

Updates include:

1. The SYNC_CONFIG_FEATURE_DISABLED check now uses FeatureUtils.validateFeature
2. Now when keys are referenced, it pulls the key using the corresponding getter function added in the Keys contract
3. Moved unassigned variable declarations back into the for loop
4. Using more easily readable syntax as is done in other contracts, for example declaring bool before if statement and referencing the bool in the if statement rather than doing all the computation within the if statement
5. Renamed cachedLatestUpdate to latestUpdateId to clarify that it is referring to the update ID and is not caching a local storage variable but retrieving the value from the DataStore contract
6. Now using the DataStore contract directly to set SYNC_CONFIG_UPDATE_COMPLETED and SYNC_CONFIG_LATEST_UPDATE_ID rather than via Config
7. Updated to use the 3 added keys for disabling updates for a given market, parameter type or market parameter combination rather than using SYNC_CONFIG_FEATURE_DISABLED as base key

SentinelDev1 committed:
Updated format to better align with other contracts in codebase

Updated the constructor and sync function format to better match other contracts in the codebase.

Given the relative simplicity of the contract and now that it is nearly complete, removed some of the more trivial comments and shortened others to try and align better with the rest of the codebase.

SentinelDev1 committed:
Update Cast.sol

Removed comments and moved variable declaration from the return statement into the function to better align with codebase

SentinelDev1 committed:
Update Keys.sol

Changed parameterType/PARAMETER_TYPE to parameter/PARAMETER and adjusted comments accordingly

SentinelDev1 committed:
Update Errors.sol

Changed ParameterType(s) to Parameter(s) and removed error SyncConfigUpdateAlreadyApplied

SentinelDev1 committed:
Update Config.sol

Changed Keys.SYNC_CONFIG_PARAMETER_TYPE_DISABLED to Keys.SYNC_CONFIG_PARAMETER_DISABLED

SentinelDev1 committed:
Update ConfigSyncer.sol

1. Changed all instances of parameterType(s) to parameter(s)

2. Added an event that is emitted when an update is completed (or if is already completed but not disabled) with the following params: uint256 updateId, uint256 prevValue, uint256 nextValue and bool updateApplied

3. Changed so that if an update has already been completed there is no error, it just skips updating for that market and parameter combination and this is reflected in the emitted event updateApplied bool (false in this case)

4. Retrieve the prevValue for the purpose of emitting the event

SentinelDev1 committed:
Update Cast.sol

Added comments back related to the bytesToUint256 function

SentinelDev1 committed:
Update Keys.sol

Added getFullKey function to Keys now that it will be used not only by the Config contract but also ConfigSyncer. Besides the slight name change, it is identical to the old _getFullKey function that used to be in Config.

SentinelDev1 committed:
Update Config.sol

Removed the _getFullKey function from this contract now that it exists (as getFullKey) in Keys

SentinelDev1 committed:
Update Config.sol

Also in prior commit, updated to use Keys.getFullKey instead of the removed _getFullKey function. This commit is just removing a blank line that was accidentally added at the end of the file.

SentinelDev1 committed:
Update ConfigSyncer.sol

Updated to use the getFullKey function in Keys rather than the ternary operator previously used.

SentinelDev1 committed:
Update and rename MockRiskOracle.sol to RiskOracle.sol

Renamed to RiskOracle as I am using the mock ReferralStorage.sol contract as an example, but can change this back if preferred.

Also updated mock RiskOracle contract to remove input params from the constructor. For initialAuthorizedSenders it's using msg.sender and for initialParameterTypes it's using a hardcoded list of all the current market parameter types.

SentinelDev1 committed:
Create deployConfigSyncer.ts

Initial deployConfigSyncer script

SentinelDev1 committed:
Create deployRiskOracle.ts

Initial version of the deployment script for the mock RiskOracle contract. 

Only deploys for hardhat, is skipped for other environments and the address specified per network in config/overwrite.ts is used instead in the deployConfigSyncer script for the RiskOracle address.

SentinelDev1 committed:
Update overwrite.ts

Added logic to handle mock RiskOracle contract address overwrite for when deploying to networks other than hardhat.

SentinelDev1 committed:
Update roles.ts

Added LIMITED_CONFIG_KEEPER role assignment on hardhat network for CONFIG_SYNCER testing

SentinelDev1 committed:
Update ConfigSyncer.sol

Changed the market param for RiskOracle.getLatestUpdateByParameterAndMarket from abi.encode(market) to abi.encodePacked(market) as the example transaction shared by Chaos Labs from their arbitrum sepolia test transaction showed the market stored as unpadded bytes, thus abi.encode returns no results.

SentinelDev1 committed:
Update and rename RiskOracle.sol to MockRiskOracle.sol

Renamed file and contract name to MockRiskOracle as I'm no longer using ReferralStorage as an example and the deployment scripts are being updated to handle this accordingly.

SentinelDev1 committed:
Update and rename deployRiskOracle.ts to deployMockRiskOracle.ts

Updated file name and contract name reference back to MockRiskOracle from RiskOracle

SentinelDev1 committed:
Update deployConfigSyncer.ts

Updated to use MockRiskOracle when network is hardhat and the deployed RiskOracle address in other scenarios. Logic will now be in a config/riskOracle.ts file instead of config/overwrite.ts

SentinelDev1 committed:
Update overwrite.ts

Reversed recent changes as the config for RiskOracle will have it's own file config/riskOracle.ts

SentinelDev1 committed:
Update index.ts

Added riskOracleConfig from config/riskOracle.ts

SentinelDev1 committed:
Create riskOracle.ts

Provides the deployed RiskOracle address when network is not hardhat

SentinelDev1 committed:
Update config.ts

Added SYNC_CONFIG_UPDATE_COMPLETED and SYNC_CONFIG_LATEST_UPDATE_ID to EXCLUDED_CONFIG_KEYS so that the "allows required keys" test for Config will pass as it should.

SentinelDev1 committed:
Update keys.ts

Updated for ConfigSyncer keys and related getter functions

SentinelDev1 committed:
Update fixture.ts

Updated to include ConfigSyncer and MockRiskOracle contracts

SentinelDev1 committed:
Create configSyncer.ts

Added for methods that will be reused for ConfigSyncer tests

SentinelDev1 committed:
Create ConfigSyncer.ts

ConfigSyncer test script, is still a work in progress but the first few tests have been added.

SentinelDev1 committed:
Update ConfigSyncer.ts

Updating indenting/formatting to better match other scripts and removed unnecessary comment

SentinelDev1 committed:
Update configSyncer.ts

Updated indenting to better align with other scripts

SentinelDev1 committed:
Update configSyncer.ts

Corrected operator in if/else if statements from "=" to "===" and changed function name computeData to getDataForKey.

SentinelDev1 committed:
Update ConfigSyncer.ts

Removed unused components from import statements and updated computeData to getDataForKey to account for function being renamed

SentinelDev1 committed:
Update ConfigSyncer.ts

Updated ethUsdMarket references from ethUsdMarket[0], ethUsdMarket[2] and ethUsdMarket[3] to ethUsdMarket.marketToken, ethUsdMarket.longToken and ethUsdMarket.shortToken respectively.

SentinelDev1 committed:
Update ConfigSyncer.sol

Added allowedBaseKeys mapping, _initAllowedBaseKeys function and _validateKey function similar to how it is done in Config.sol (except _validateKey does not validating roles in this contract). This makes it so only parameters with specific baseKeys can be updated via ConfigSyncer and prevents updates for parameters with other baseKeys that aren't explicitly whitelisted.

SentinelDev1 committed:
Update configSyncer.ts

Updated test parameters to include 5 that are whitelisted and 1 that is not. Now including parameters that test for when longToken and shortToken are applicable and for when isLong = true and isLong = false.

SentinelDev1 committed:
Update ConfigSyncer.ts

Added more tests, removed unused imports and variables and added a few comments that will be removed once complete.

SentinelDev1 committed:
Update ConfigSyncer.ts

Reorganized the imports and added an additional comment around changes I made to the updated parameter values assigned in beforeEach

SentinelDev1 committed:
Update ConfigSyncer.sol

Added additional keys to _initAllowedBaseKeys

SentinelDev1 committed:
Update ConfigSyncer.ts

Added more tests for validating that the sync function will revert with the correct error when updates are disabled for a given market, parameter or market parameter combination, when the SyncConfig feature is disabled or a provided parameter is not whitelisted

SentinelDev1 committed:
Update ConfigSyncer.sol

Changed local variable name syncConfigUpdatedCompleted to syncConfigUpdateCompleted and removed the abi.encodePacked conversion of market to bytes for the RiskOracle getLatestUpdateByParameterAndMarket() function call in the sync function to account for market being updated in the RiskOracle contract from bytes to address.

SentinelDev1 committed:
Update IRiskOracle.sol

Changed variable 'market' from bytes to address to align with update to the RiskOracle contract

SentinelDev1 committed:
Update MockRiskOracle.sol

Updated variable 'market' from bytes to address to align with the change made to the Chaos Labs RiskOracle contract.

SentinelDev1 committed:
Update Errors.sol

Added new SyncConfigInvalidMarketFromData error for instances where in ConfigSyncer, the provided market address doesn't match the market address derived from additionalData

SentinelDev1 committed:
Update configSyncer.ts

Updated updateFormat to parameterFormat given it is related to parameters.

SentinelDev1 committed:
Update ConfigSyncer.ts

Added test for instances where the ConfigSyncer market address doesn't match the market address derived from the additionalData retrieved from the RiskOracle contract.

SentinelDev1 committed:
Update ConfigSyncer.sol

Added _validateMarketInData function to validate that the market address param input in sync is equal to the market address derived from the additionalData retrieved from the RiskOracle contract.

Added RESERVE_FACTOR and OPEN_INTEREST_RESERVE_FACTOR keys to allowedBaseKeys mapping.

xvi10 commented on contracts/config/ConfigSyncer.sol:
```diff
@@ -0,0 +1,173 @@
+        address marketFromData;
+        if (baseKey == Keys.MAX_PNL_FACTOR) {
+            bytes32 unusedBytes32;
+            bool unusedBool;
```
i don't think these are needed, could just comment out the unused variables, example: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/token/TokenUtils.sol#L62

SentinelDev1 replied:
I've updated per the example provided in the latest commit. I also changed the local variable naming from unusedBytes32 and unusedBool to extKey and isLong respectively, for more informative naming. 

Variable name 'extKey' stands for extension or extended key, which is the naming I created and used in utils/configSyncer.ts for the additional key used for parameters with baseKey = MAX_PNL_FACTOR as these were unnamed in the updateMarketConfigUtils.ts file I derived them from and I wanted to differentiate them from baseKey to avoid confusion.

If preferred, I can revert this naming change.

SentinelDev1 replied:
I think this is resolved, but I will let you confirm on your end in case I am somehow missing something here.

SentinelDev1 committed:
Update ConfigSyncer.sol

Renamed unusedBytes32 and unusedBool to extKey and isLong respectively for more informative naming.

Moved these local variable declarations into the abi.decode and commented out them out given that they are not used.

SentinelDev1 committed:
Update toggleFeatures.ts

Added SYNC_CONFIG_FEATURE_DISABLED (commented out for now similar to others)

SentinelDev1 committed:
Update riskOracle.ts

Updated addresses for final Chaos Labs RiskOracle contract deployments to Arbitrum Sepolia and Avalanche Fuji.

SentinelDev1 committed:
Update MockRiskOracle.sol

Updated using final Chaos Labs RiskOracle contract. 

Previously, the constructor did not take inputs and instead had authorizedSenders[msg.sender] = true and the initialUpdateTypes[] hardcoded to include 50 parameters retrieved from updateMarketConfigUtils.ts but now the initialSenders and initialUpdateTypes are included as constructor input parameters and passed to the contract from the deployMockRiskOracle.ts script on deployment. 

The initialUpdateTypes has also been updated to align with the allowedBaseKeys included in the ConfigSyncer contract, plus maxPnlFactorForTradersLongs which is used in the ConfigSyncer.ts test script.

Please note that this updated MockRiskOracle contract is different from the final Chaos Labs RiskOracle contract in the following ways:

1. Changed to pragma solidity ^0.8.0 from pragma solidity ^0.8.25 because the latest version available in the repo is 0.8.18 and set to 0.8.0 to align with what is used for other contracts in the repo

2. Removed Ownable(msg.sender) from constructor as it appears the outdated version of the OpenZeppelin Ownable contract in the gmx-synthetics repo doesn't take this constructor input and just directly assigns Owner to msg.sender within the constructor

3. Changed the contract name from RiskOracle to MockRiskOracle

SentinelDev1 committed:
Update deployMockRiskOracle.ts

Updated to pass 'initialSenders' and 'initialUpdateTypes' as constructor param inputs to the updated MockRiskOracle contract on deployment.

The 'initialSenders' array is set to include just the deploying 'wallet' address used for tests in the ConfigSyncer.ts test script.

The 'initialUpdateTypes' array is set to include all parameters with base keys that equal true in the ConfigSyncer contract allowedBaseKeys plus maxPnlFactorForTradersLongs which is used for tests in the ConfigSyncer.ts test script. This should fix the two remediation findings from the Guardian audit (L-01 and L-02).

SentinelDev1 committed:
Update MockRiskOracle.sol

Updated per the latest version of the RiskOracle contract. The main change is the validation Chaos Labs added in the constructor that prevents pushing duplicate update types to the allUpdateTypes array from the constructor input param 'initialUpdateTypes' during deployment.

SentinelDev1 committed:
Update riskOracle.ts

Updated for the latest RiskOracle arbitrum sepolia and avalanche fuji contract addresses

SentinelDev1 committed:
Create updateRiskOracleConfig.ts

Initial draft script for disabling/re-enabling specific market, parameter and/or market parameter combinations.

Can make more dynamic if needed.
